storage/ndb/src/kernel/vm/mt.cpp

```diff
@@ -1228,65 +1228,82 @@ calc_fifo_free(Uint32 ri, Uint32 wi, Uint32 sz)
 /**
  * Identify type of thread.
  * Based on assumption that threads are allocated in the order:
- *  main, ldm, query, recover, tc, recv, send
+ *  ldm, tc, recv, main, rep
  */
+
 static bool
 is_main_thread(unsigned thr_no)
 {
-  if (globalData.ndbMtMainThreads > 0)
-    return (thr_no < globalData.ndbMtMainThreads);
-  unsigned first_recv_thread = globalData.ndbMtLqhThreads +
-                               globalData.ndbMtRecoverThreads +
-                               globalData.ndbMtTcThreads;
-  return (thr_no == first_recv_thread);
+  if (globalData.ndbMtMainThreads == 0)
+    return (thr_no == g_first_receiver_thread_no);
+  else
+    return (thr_no ==
+     (g_first_receiver_thread_no + globalData.ndbMtReceiveThreads));
 }
 
 static bool
-is_ldm_thread(unsigned thr_no)
+is_rep_thread(unsigned thr_no)
 {
-  if (glob_num_threads == 1)
-    return (thr_no == 0);
-  return thr_no >= globalData.ndbMtMainThreads &&
-         thr_no <  globalData.ndbMtMainThreads + globalData.ndbMtLqhThreads;
+  if (globalData.ndbMtMainThreads == 0)
+  {
+    if (globalData.ndbMtReceiveThreads == 1)
+    {
+      return (thr_no == g_first_receiver_thread_no);
+    }
+    else
+    {
+      return (thr_no == (g_first_receiver_thread_no + 1));
+    }
+  }
+  else
+  {
+    unsigned main_base = g_first_receiver_thread_no +
+                         globalData.ndbMtReceiveThreads;
+    if (globalData.ndbMtMainThreads == 1)
+    {
+      return is_main_thread(thr_no);
+    }
+    return (thr_no == main_base + 1);
+  }
 }
 
 static bool
-is_recover_thread(unsigned thr_no)
+is_main_or_rep_thread(unsigned thr_no)
 {
-  Uint32 num_recover_threads = globalData.ndbMtRecoverThreads;
-  unsigned query_base = globalData.ndbMtMainThreads +
-                        globalData.ndbMtLqhThreads;
-  return thr_no >= query_base &&
-         thr_no <  query_base + num_recover_threads;
+  if (is_main_thread(thr_no) || is_rep_thread(thr_no))
+    return true;
+  return false;
 }
-
-bool
-mt_is_recover_thread(Uint32 thr_no)
+static bool
+is_ldm_thread(unsigned thr_no)
 {
-  return is_recover_thread(thr_no);
+  if (globalData.ndbMtLqhThreads > 0)
+  {
+    return (thr_no < globalData.ndbMtLqhThreads);
+  }
+  else
+  {
+    return (thr_no < globalData.ndbMtReceiveThreads);
```

Comment:
It seems that if MtlqhThreads == 0, then we assume that lqh threads are sharing thread with both Tc and Receiver threads?

Is there any possibility that Lqh only shares Tc threads?

Reply:
ok, yep, if TcThreads > 0 then also LdmThreads > 0